### DISCUSSION TITLE
Ideas for LocaleSpecification

### DISCUSSION THREAD

**Discussion by pshinners-blz** [OpenAssetIO/OpenAssetIO#229] [Ideas and change proposals]
When integrating our asset system with Katana, there were many conditions we wanted to know why the host application was calling into the assets. Based on information from recent WG meetings, it sounds like this may arrive in the form of LocaleSpecifications. 

I've looked over our code and evaluated what data would be useful. This could be the starting point for what data is defined in this specification. For each type of data I found I'll try to note why its useful, but also identify potential alternatives.

1.*Host info* First of all, just knowing the name of the application and version number would be helpful any time I am looking into errors or tracing from an OpenAssetAPI plugin. Perhaps it is decided it would be wiser to glean this information from the running process? Would passing a few informational strings into every single asset call come with notable overhead? Since this data seems static, it could be provided through environment variables?
2. *Read or Write* Many of the resolve functions want to know if the system is resolving an asset id for writing or for reading. Perhaps this should be implemented as separate calls the asset plugin implements. Not every asset call would actually be affected by this.
3. "user info* The documentation for Katana's "Check permissions" function has examples of a stringmap including information like "username" and "workstation". Would this be useful generic information for all calls? I could imagine a fine-grained permission model using some form of credentials for resolving asset ids? It seems hard for a host application and asset plugin to ever agree on what data is used for credentials. I don't think Katana's "check live group permissions" actually ever implemented any data like this, it was purely theoritical?
4. Our asset plugin tries to infer user actions like "File Save", "File Save As", "File Version Up", and "Live Group Publish" by analyzing the asset type arguments and extra "versionUp" and "publish" arguments. It seems so much cleaner to pass a name for whatever operation triggered the asset api. Unfortunately, knowing the "name of a user operation" can be really hard (I'm thinking nested undo groups, etc)
5. Katana's parameter ui delegate has separate functions for `createAssetControlWidget` and `createAssetRenderWidget`. This seems like data better suited to a LocaleSpecification. While at it, Katana also defines other asset parameter types like `assetIdInput`, `assetIdDirectory`, and these all can map to different parameter widgets like `assetIdOutput`, `fileInput`, or `liveGroupSource`. Perhaps constructing widgets for values is a very host-specific operation and not related to openassetio? I could also imagine a world where the host applications provide rich asset widgets built in, that cooperate extensively with the openassetio plugin, and the widgets do not get replaced by custom delegates?
6. Our pipeline generally does not support "implicit versions" (like using 'best' or 'lastest'). But our asset plugin does need to support this for Katana's recent file menu, so we have special exceptions to when strict versions are required in the asset id resolver based on the asset type and our guess as to why Katana is resolving an implicit asset id. This may fall into the same category as #4, some reason why Katana is callinging into the asset plugin so we can avoid heuristic behavior in our plugin.

After looking through our use cases and coming up with this data, I'm not convinced LocaleSpecifications are a requirement. If the API for writing asset plugins was designed to make information like this known through other means, then that sounds potentially cleaner.

**Comment by pshinners-blz** [comment]
Oh, I forgot one
1. Name of node and/or parameter creating the request. This mostly useful for tracking things like "which render node generated this image" or "which bake node generated this katana look". Besides just logging this information to the asset system, I could see us doing validations and data tracking. In fact we currently do some of this in our Katana UI delegate, since that gets lots of information about what initiated the browser for publishing. But we don't get any of that data during a non-interactive publish.

Looking back on these ideas, most of it is really only interesting in the context of writing. Perhaps instead of `LocaleSpecification` this should be considered `WriteSpecification` and only passed into write calls in the asset plugin. But perhaps this ends up putting a burdon on host applications that must now fill out a form and twenty questions just to run "file save"?

**Comment by foundrytom** [comment]
> I've looked over our code and evaluated what data would be useful. This could be the starting point for what data is defined in this specification. For each type of data I found I'll try to note why its useful, but also identify potential alternatives.

Thanks for all the info here @pshinners, this is most valuable and raises some excellent topics. Notes on how things currently stand:

> 1._Host info_ First of all, just knowing the name of the application and version number would be helpful any time I am looking into errors or tracing from an OpenAssetAPI plugin. Perhaps it is decided it would be wiser to glean this information from the running process? Would passing a few informational strings into every single asset call come with notable overhead? Since this data seems static, it could be provided through environment variables?

We wanted to try and make as much of the "context that could affect behaviour" as explicit as possible to try and reduce the margin for error via side channels. As part of the 'stateless' nature of the `ManagerInterfrace` API, each method is passed a [`HostSession`](https://github.com/TheFoundryVisionmongers/OpenAssetIO/blob/main/python/openassetio/managerAPI/HostSession.py). This is used for logging, and also provides the implementation access to an instance of the [`Host`](https://github.com/TheFoundryVisionmongers/OpenAssetIO/blob/main/python/openassetio/managerAPI/Host.py) interface. Which you can use to stably identify the host. As its a query, it doesn't have any overhead per-call data-wise.

> _Read or Write_ Many of the resolve functions want to know if the system is resolving an asset id for writing or for reading. Perhaps this should be implemented as separate calls the asset plugin implements. Not every asset call would actually be affected by this. 

Our approach to this is the [`Context`](https://github.com/TheFoundryVisionmongers/OpenAssetIO/blob/main/python/openassetio/Context.py) object passed to each API call.  The supplied context's `access` is used to denote read or write.  The context also contains the `locale` (as you correctly guessed - a `LocaleSpecification`) and a few other bits of information.

We hoped having this as input to a method vs new methods would make it easier for a manager to decide if it needed to adapt its behaviour, without making it seem like a requirement of the API itself. It also means that the context object is always _the_ way that this is communicated across different methods (eg: `managementPolicy`). So I guess as it stands, the API is partitioned by activity, and parameterised by context which felt like a reasonable split that also helps avoid the interface getting too large.

>. "user info* The documentation for Katana's "Check permissions" function has examples of a stringmap including information like "username" and "workstation". Would this be useful generic information for all calls? I could imagine a fine-grained permission model using some form of credentials for resolving asset ids? It seems hard for a host application and asset plugin to ever agree on what data is used for credentials. I don't think Katana's "check live group permissions" actually ever implemented any data like this, it was purely theoritical? 

This is a really interesting topic. We forked the Katana asset API's conceptual model before permissions were added. We have #9, to track that we need to add this to OpenAssetIO before a "v1". It'd be great to dig into this in more detail then. 

When we recently spoke to Movielabs last about how this overlapped with their views on the security model of the future, it seemed the "assume you are running within a trusted process environment/walled garden where identity is confirmed" strategy might mean some of this could be handled indirectly through that environment rather than the API, to avoid the situation you described, where it's hard to agree on what's needed! It'd be great to look into this into more detail. 

> 4. Our asset plugin tries to infer user actions like "File Save", "File Save As", "File Version Up", and "Live Group Publish" by analyzing the asset type arguments and extra "versionUp" and "publish" arguments. It seems so much cleaner to pass a name for whatever operation triggered the asset api. Unfortunately, knowing the "name of a user operation" can be really hard (I'm thinking nested undo groups, etc) 

In our working implementation, this is exactly the sort of thing we'd have in the `context.local` `LocaleSpecification` so you'd have locales roughly like (vague pseudo code): `locale:action.file.save`/`locale:action.file.saveAs`/`locale:action.file.publish`. The big debate-to-come is then working out the most sensible hierarchies for these specifications, but hopefully the underlaying mechanism makes sense, and allows some degree of generic behaviours to be built into the manager in an application-agnostic way if required.

>  Katana's parameter ui delegate has separate functions for `createAssetControlWidget` and `createAssetRenderWidget`. This seems like data better suited to a LocaleSpecification. 

Yeah, in our PoC, we used an identifier to define a widget that is needed, so there is just one `getWidget( identifier, ... )` method. (We haven't ported the UI layer from the `FnAssetAPI` prototype yet), this is prob Q2 this year. This may also make sense to be a spec instead too. I guess it depends on how parameterised it needs to be.

> While at it, Katana also defines other asset parameter types like `assetIdInput`, `assetIdDirectory`, and these all can map to different parameter widgets like `assetIdOutput`, `fileInput`, or `liveGroupSource`. Perhaps constructing widgets for values is a very host-specific operation and not related to openassetio? I could also imagine a world where the host applications provide rich asset widgets built in, that cooperate extensively with the openassetio plugin, and the widgets do not get replaced by custom delegates?

Yeah, this sounds very much along the lines we were thinking.  In the PoC, we had the application assemble the final UI from a mixture of its own components, and manager specific ones where required for or provided, so the user has a more coherent experience. 

When we get to this in the coming months, it'd be great to delve into this whole topic properly.

>  6. Our pipeline generally does not support "implicit versions" (like using 'best' or 'lastest'). But our asset plugin does need to support this for Katana's recent file menu, so we have special exceptions to when strict versions are required in the asset id resolver based on the asset type and our guess as to why Katana is resolving an implicit asset id. This may fall into the same category as 4, some reason why Katana is callinging into the asset plugin so we can avoid heuristic behavior in our plugin.

It'd be good to dig into this in more detail at some point (particularly when we do #10) - as our current thinking is that a host should never have any explicit understanding about implicit vs explicit versions - and it be totally up to the manager to decide if this is a supported/applicable concept. In OpenAssetIO in general, the host is required to treat more things as opaque and give control to the manager, which may mean the delegation of larger parts of the UI where interactions with any specific semantics of how entities are structured or managed is required.

> After looking through our use cases and coming up with this data, I'm not convinced LocaleSpecifications are a requirement. If the API for writing asset plugins was designed to make information like this known through other means, then that sounds potentially cleaner.

It would certainly simplify things if they're not. Do you have any idea what these "other means" could be? We had been thinking this ourselves recently, but then we came up with quite an interesting case for them whilst looking at #215. 

We have seen several cases in the wild where people want to have "vague" entity references that would resolve in a different way depending on the context of resolution. A contrived example being: `asset:///slateBG?v=latest` or `asset:///shotCamera?v=latestApproved`.  This is to allow a generic template to be used across multiple shots, etc...  There are a plethora of ways of doing this, for example, using string substitution based on the environment, graph state variables etc. host-side to form an actual concrete entity reference. But the idea of references that have some context variability have come up several times, and there may well be many cases where a host doesn't have a suitable substitution mechanism of its own.

In the case of #215, the `OpenTimelineIOLocale` could have a `metadata` property that contained the flattened inherited OTIO metadata for each clip, which could contain the variables that would have been used in the host-side substitution.  A locale in katana could have `graphStateVariables` etc...

This would allow that context-specific variability to be handled by the manager in a consistent way. Meaning OpenAssetIO itself can remain a 'pass through', rather than modifying the references at all, without relying on the environment, which doesn't work in "multi-shot" (and similar) contexts.

What are your thoughts on these kind of references? I've heard support for both sides of the argument - the one that says entity references should be concrete and this should be a string substitution challenge before the manager gets involved, but then also arguments to the contrary to allow more scope for a pipeline to define context-sensitive entity references where the manager implementation takes care of this kind of dynamic behaviour.

I think well build this capability into #215 as an experiment to at least facilitate a discussion around real-world use cases.

> Name of node and/or parameter creating the request. This mostly useful for tracking things like "which render node generated this image" or "which bake node generated this katana look". Besides just logging this information to the asset system, I could see us doing validations and data tracking.

This was definitely the thing we wanted to make simple with the context `locale` being a spec. Where you'd have, for (contrived) example: `locale:nodegraph.node.writer` locale spec, with a `{ nodeID: <someNameOrHandle>, ... }`, and then using the `Host` interface supplied to the method, you'd know which app, we some half formed ideas (#7) around some kind of method to return you the native SDK object for `<someNameOrHandle>` (as we can't pass them directly through the locale for various reasons).

 > In fact we currently do some of this in our Katana UI delegate, since that gets lots of information about what initiated the browser for publishing. But we don't get any of that data during a non-interactive publish.
Looking back on these ideas, most of it is really only interesting in the context of writing.  Perhaps instead of LocaleSpecification this should be considered WriteSpecification and only passed into write calls in the asset plugin.

When we were putting all this together, having the same mechanism being used for policy, browsing and publishing felt like it made things a little more consistent. We're hoping to have some more concrete and real-world end-to-end examples soon, so I'd be great to see what you think of how it works out in practice as its definitely something we want to try and get right.

> But perhaps this ends up putting a burdon on host applications that must now fill out a form and twenty questions just to run "file save"?

We saw so many cases where the manager was trying to reverse engineer this info, it felt reasonable to try and get the host to participate in this a little more, but its definitely a balance to strike. This is where the info you've given here is really helpful as it helps steer what we (collectively) think should be "required" from the host to be a good team player.  Im hoping we can find a good compromise that makes life easier for the manager without being a p.i.t.a for the host.

**Comment by pshinners-blz** [comment]
Taking a high level look at our Katana asset plugin and publish tools we've built into it, it feels like there are two main modes for writing to the asset system.

1. "Write new version"
2. "Write into/over something that exists"

When I mentioned the LocaleSpecification might not be so necessary I was wondering if separate write methods would simplify trying to map both these use cases to the same write and resolve functions.

Other studios may use other types of writing, but I suspect there is a small, finite list of different styles of exporting data from applications.

Our current Katana asset plugin distinguishes between these two writing modes by the asset id provided. It either has an explicit version number as part of the id to write into a specific version that has already been defined. If the asset id has no specific version then the write goes to a newly created version by the publish. The Parameter widgets and our custom browser delegate steer the user towards either of these types of actions.

But when the action is initiated by Katana without one of our widget delegates, thats when we are having to figure out what Katana intends, by inspecting the "publish" and "versionUp" fields in the additional arguments.

**Reply by foundrytom** [reply]
Thanks @pshinners this would be a good one to dig into a more little soon. One of the conceptual challenges we encountered before is whether the caller always knows enough information to be able to pick the right method (especially for write). There were many cases where it didn't, so we ended up with a pattern similar to the Katana one, just with much more concrete context about the caller. So hopefully, you would know enough about what is happening though the suitably configured `LocaleSpecification`.

**Comment by pshinners-blz** [comment]
Allowing "relative" versus "absolute" asset ids seems super valuable but we haven't taken advantage of it ourselves.

One place I see something similar in existing applications is Houdini's file browser. It has a checkbox that lets the artist control if the returned file path has environment variables resolved or not. Since many studios have variables for things like `$SHOT`, it means the browser can choose if a file reference is relative or absolute. (Sort of, none of that solves picking a version on different shots...)

In the end, our Katana asset plugin requires absolute asset references, because there's nowhere to provide a current context to the asset plugin. (We do have gsv variables that define where the artist is working used by our ui delegates). Our custom parameter widgets in Katana allow users to enter relative asset ids, but the UI resolves them before applying values to the parameters. This is helpful for many cases, but does not allow "multishot" nodes to work across different contexts. That sounds useful, but it's a road we haven't explored for production yet.
